Pending ADR for new approach to monitoring and alerting (#1476)
alphagov/email-alert-api: docs/adr/adr-007-monitoring-and-alerting.md

Notify is working as they are independently monitored. The use of
"if and only if" reflects a condition we're forming that the system will not
raise critical alerts unless we're highly confident it is broken, whereas
warnings are appropriate for other situations. We decided that for the email
alert system to be "broken" it represents a failing that will lead to a user
noticeable issue unless an engineer intervenes. Since these alerts should
represent a broken system requiring intervention, it should be suitable to
contact engineers out-of-hours for them as otherwise users will be
experiencing issues.

We decided that we will re-approach our alerting with this mindset and look to
replace alerts that are vulnerable to reflecting a busy status.

## Status

Pending

## Consequences

We will be changing the approaching to monitoring and alerting across the email
alert system of applications in order to meet this new alerting criteria.

## Monitoring Email Alert API

### Sidekiq workers

We will configure critical alerts, that contact engineers out-of-hours, if
individual Sidekiq jobs fail multiple times for the following workers:

- [DailyDigestInitiatorWorker][]

> - [DigestEmailGenerationWorker][]

Reviewer: Thanks, the new content looks good, except for the next paragraph, which starts off with "Similarly" - I get that it's a similar alert, but the reasoning is (now) quite different. Minor, but could we just remove "Similarly"?

- [ProcessContentChangeWorker][]
- [ProcessMessageWorker][]

> - [SendEmailWorker][]

Reviewer: Sure, that's OK. I was trying to remember it's behaviour regarding rate limiting, which lead me to suggest this. In general, I think we're asserting that none of these jobs should be prone to failure, so we can treat them all equally.

Does that match your understanding?

> - [SendEmailWorker][]

Author: Yup, rate limiting doesn't result in a failure (the worker is [rescheduled](https://github.com/alphagov/email-alert-api/blob/1711885b4df9c80aed5d8d2333e4e007c778badd/app/workers/send_email_worker.rb#L16-L17)). `temporary-failure` is also something we hear about in a call back rather than at sending time.

Yeah, from the doc our decision was to treat these the same. With the external service there is a greater risk of a problem in SendEmailWorker but I recall us concluding the odds of 5 intermittent fails in a row were still absurdly low.

- [WeeklyDigestInitiatorWorker][]

These workers will use the database to store the amount of times we have
attempted to perform an action. An alert will be raised if the quantity of
failed attempts reaches a threshold (the suggestion is 5).

We will contact engineers out-of-hours if either [SendEmailWorker][] or
[DigestEmailGenerationWorker][] have not completed at least one successful run

> within a time period (suggestion is 30 minutes) despite work existing. This

Reviewer: Sure, I can explain. Currently our pattern is to only recover lost work [after an hour](https://github.com/alphagov/email-alert-api/blob/469efe788a290d1a2205d8f870c6aafb4c8c9b00/app/workers/recover_lost_jobs_worker/unprocessed_check.rb#L13). Even if we reduced the that to 30 mins, the worker only runs every half hour, so it could be up to an hour before we recover a lost email. If the total number of emails being sent is small in total (let's say we're processing a "tiny" content change at 11PM), then in the worst case it would be (say) 59 mins between the emails being created and a lost one being recovered and sent.

There's always the possibility of being unlucky, of course. Since we know this happens every so often, I was wondering if we should mention it here as a general caveat: the thresholds may need to be longer (or the recovery more frequent) to account for one or two losses in work.

Does that help?

> within a time period (suggestion is 30 minutes) despite work existing. This

Author: Thanks, I understand now and is a valid point. 

The time period in RecoverLostJobsWorker should ideally be less than the period before we alert and really shouldn't be greater than it. I've modified the ADR to include a reference to this but haven't gone the whole way in defining the solution now - hopefully we can that out when working out how to retry SendEmailWorker (and apply the same to the DigestEmailGenerationWorker).

> within a time period (suggestion is 30 minutes) despite work existing. This

Reviewer: Yes, I agree we can sort this out as an implementation detail.

if all instances of these jobs are failing and the aforementioned retry method
may take a long time to alert in a busy period.

Similarly, we will contact engineers out-of-hours if we determine that
[DigestRuns][DigestRun] are not created within a reasonable time frame
(suggestion is 2 hours). Creating these is a [trivial][digest-run-creation],
but essential, operation for producing a digest and the lack of these suggests
the system is broken whether it is busy or not.

We will also monitor that runs of [RecoverLostJobsWorker][] and
[MetricsCollectionWorker][] are succeeding at least once within a time frame
(suggestion is 1 hour). These jobs should continue running even when the
system is busy and the lack of them running indicates that we no longer have
confidence the system is monitored or able to recover. Thus, an alert will be
created that contacts engineers out-of-hours should these never succeed in a
time frame.

For Sidekiq workers that don't affect the user noticeable aspects of Email
Alert API, such as [EmailDeletionWorker][], we won't raise critical alerts if
these are failing. Instead we will raise a warning if any Sidekiq workers

> exhaust all their retries and are placed in the [dead set][sidekiq-dead]. This

Reviewer: Minor: I think it's worth mentioning (at a high level) that the retries should be adjusted so they are exhausted in a short-ish period, as otherwise someone reading may wonder when exactly this alert would go off.

Reviewer: Thanks for updating 👍.

application becoming unavailable - unless travel advice were published. We
noted there wasn't a particularly consistent approach to this across GOV.UK
([with some apps raising an out-of-hours alert when a single instance fails
healthchecks][publishing-api-alert]) and felt it would be good to take a simple
approach to avoid surprising engineers.

We decided the most appropriate place to alert would be when [the application
load balancer has no healthy instances of the application][aws-lb-check]. This
represents that no instances of the web application are healthy and
[no requests will succeed][aws-alb-health].

### Alerts to be removed

We will remove the following alerts that will become superseded:

- [Content Changes unprocessed after 2 hours][unprocessed-content-changes]
- [Messages unprocessed after 2 hours][unprocessed-messages]
- [Incomplete DigestRuns after 2 hours][incomplete-digest-runs]

## Monitoring Email Alert Service

Email Alert Service listens for [Publishing API][] events and communicates
these to Email Alert API. We currently [only have alerts][rabbitmq-alert] that
occur in-hours if this system fails, so there is no visibility for any
out-of-hours problems.

> We intend to improve this by raising an alert that contacts engineers

Reviewer: ```suggestion
We intend to improve and consolidate the existing alerts, and contact an engineer
```

Minor: thought it worth a mention that we're not intending to increase the number of RabbitMQ alerts.

out-of-hours if there is a pattern of failure in processing these events.

## Monitoring Email Alert Frontend

Email Alert Frontend acts as the public interface for subscriber interaction
with the email alert system. This web application will be monitored in a
[similar way to the Email Alert API web application][#web-application] by
considering whether the load balancer has any healthy instances.

## Removal of Email Alert Monitoring

[Email Alert Monitoring][email-alert-monitoring] is the aforementioned
end-to-end test that monitors whether travel advice and drug advice content
changes send email to a Gmail inbox. The changes to alerting described in this
ADR will render Email Alert Monitoring redundant, as the new alerts should
capture real problems faster and cover a broader range of scenarios - this
would leave Email Alert Monitoring capturing false positives and out-of-scope
problems.

> Once the new monitoring and alerting approaches have been implemented and

Author: Thanks I've added a section on this.

would reach a peak of sending 1 million emails a day whereas now it can be
sending over 10 million emails a day. This elevated volume of work has
increased system latency which has impacted alerts and contributed to
incidents.

The approach used to monitoring the application and the wider email alert
system is two fold. One approach is to monitor Email Alert API's internal
state to see if it is [completing tasks within a particular
time frame][monitor-time-frame] or [whether there is a high quantity of work to
do][monitor-workload]. The other approach is to use an [end-to-end
check][email-alert-monitoring] which verifies that a Gmail account
has received an email within a time frame. Both of these approaches are flawed.

Monitoring for high quantities of work, and whether work was completed in a
time frame, has left Email Alert API vulnerable to producing alerts when the
application is busy, yet fully operational. These alerts are unactionable for
support engineers as the alerts resolve themselves once the work is completed
and there is little an engineer can do to speed this process up. For example,
if the application has accumulated 1,000,000 emails to send it raises a

> critical alert even if it is sending these at the fastest rate Notify

Author: Thanks, that's a good idea.

would allow. These alerts, therefore, waste the time of support engineers and
create a risk that a support engineer might miss an alert that requires their
intervention.

Since these alerts can't be relied upon to reliably suggest the
application is broken, they aren't used for any out-of-hours alerts. Therefore,
even if they did reflect a genuine problem no engineers would be contacted.
An engineer instead learns of a problem out-of-hours due to the other type
of monitoring approach: the end-to-end checks. These alert when content
published through [Travel Advice Publisher][] does not arrive in
a particular Gmail inbox within a specified time period.

> Monitoring an inbox for a limited set of emails has its own distinct flaws.

Author: I don't think we need to try make this list exhaustive. This seems useful extra arguments if we need to work harder to convince someone of their problems, but that seems beyond the scope of this ADR.

the system was broken or not. We felt there was room for improvement.

[monitor-time-frame]: https://github.com/alphagov/email-alert-api/blob/c70fcd7e5d64393074a3eebd83b3467fd8cc03b3/app/workers/digest_run_worker.rb#L6-L9
[monitor-workload]: https://github.com/alphagov/govuk-puppet/blob/fac4b5163f10736bb19aa51a8eb7bc276dd2cb40/modules/govuk/manifests/apps/email_alert_api/sidekiq_queue_check.pp#L33-L44
[email-alert-monitoring]: https://github.com/alphagov/email-alert-monitoring
[Travel Advice Publisher]: https://github.com/alphagov/travel-advice-publisher

## Decision

We decided that the system should raise a critical alert
[if and only if](https://en.wikipedia.org/wiki/If_and_only_if) the email alert
system is broken. The majority of these alerts should contact engineers
out-of-hours.

Unpacking this, we consider the email alert system to be comprised of 3
applications: Email Alert API,
[Email Alert Frontend](https://github.com/alphagov/email-alert-frontend) and
[Email Alert Service](https://github.com/alphagov/email-alert-service); this
excludes the monitoring of whether GOV.UK publishing is working and whether

> Notify is working as they are independently monitored. The use of

Author: Thanks 👍

warnings are appropriate for other situations. We decided that for the email
alert system to be "broken" it represents a failing that will lead to a user
noticeable issue unless an engineer intervenes. Since these alerts should
represent a broken system requiring intervention, it should be suitable to
contact engineers out-of-hours for them as otherwise users will be
experiencing issues.

We decided that we will re-approach our alerting with this mindset and look to
replace alerts that are vulnerable to reflecting a busy status, or which
extend beyond the boundaries of the system.

## Status

Pending

## Consequences

We will be changing the approaching to monitoring and alerting across the email
alert system of applications in order to meet this new alerting criteria.

### Monitoring Email Alert API

#### Sidekiq workers

We will configure critical alerts, that contact engineers out-of-hours, if
individual Sidekiq jobs fail multiple times for the following workers:

- [DailyDigestInitiatorWorker][]
- [DigestEmailGenerationWorker][]
- [ProcessContentChangeWorker][]
- [ProcessMessageWorker][]
- [SendEmailWorker][]

> - [WeeklyDigestInitiatorWorker][]

Author: They're already alphabetical - the casulty of alphabetical was the Digest ones were no longer grouped, however that made the whole list ordering arbitrary.

> These workers will use the database to store the number of times we have

Author: That seems more of an implementation detail.

attempted to perform an action. An alert will be raised if the quantity of

> failed attempts reaches a threshold (the suggestion is 5). To ensure that these

Reviewer 2: Could you elaborate on why 5 was chosen? 

Does this number also need to be adjusted per worker? 5 failed attempts trying to send a single email seems minor, 5 failed attempts trying to process a daily digest seems major.

Author: I've added a brief comment to justify 5. The same would be applied to all the workers. We worked out that with intermittent network problems we'd be extremely unlucky to have the same email 5 times. All the other workers don't make external requests to services so should be very unlikely to fail.

There's an aspect from the original document of seeing if we got any SendEmailWorker false positives due to intermittent failures. Ideas on approaching it could be automatically retrying within the worker on SendEmailWorker failures or setting that worker to be more lenient. Hopefully this won't be needed though.

I'm not too sure how to add this into the ADR without it going off on a tangent... 🤔

Reviewer 2: Your brief comment does help but could we add something along the lines of your first paragraph as I think it's helpful to highlight the decision up on why 5 was chosen and applied to all workers equally?

Author: Sure, I've updated it some more.

Reviewer 2: Much clearer, thanks!

jobs actually run we will configure the [RecoverLostJobsWorker][] to identify
any of these that need to run and attempt them, this protects from lost work

> problems such as [Sidekiq crashing][sidekiq-reliability].

Reviewer 2: Is this sentence referring to the work we've already done in the RecoverLostJobsWorker, or future work?

Author: A bit of both - we've done some things and still have more to go.

created that contacts engineers out-of-hours should these not succeed in a
time frame.

For Sidekiq workers that don't affect the user noticeable aspects of Email
Alert API, such as [EmailDeletionWorker][], we won't raise critical alerts if
these are failing. Instead we will raise a warning if any Sidekiq workers
exhaust all their retries and are placed in the [dead set][sidekiq-dead]. This
approach was chosen as it was a pragmatic technique to monitor multiple workers
without adding alerts for individual workers. Workers will have their
[retry cadence][sidekiq-retry] considered so that this will alert in a
reasonable time frame of the retries being exhausted (the Sidekiq default of 21
days is very long).

We will modify the [latency checks for email queues][latency-checks]. We
will be treating [transactional emails][] different to other latency checks
as these emails have have an [expectation of a prompt
delivery][transactional-delivery-expectation] and the lack of them blocks a
user from progressing with their task. These will raise a warning after 5 mins
and go critical (with engineers contacted out-of-hours) should the latency

> reach 15 mins.

Reviewer 2: Are we planning to call up people if this only applies to a single transactional email which is not sending?

Author: It's based on the queue latency and doesn't consider the number of items. In theory it could be possible to have one item in the queue and a latency of 15 minutes but in practice that should be virtually impossible as you'd only have queue latency when there is a high number of items in the queue.

Reviewer 2: My thoughts were around as we have the `send_email_immediate_high` queue which has the same priority as as the transactional queue then could it be feasible that a single transactional email gets stuck behind a load of high priority emails?

Author: Oh I had thought the transactional email was a higher priority than immediate high, that is a surprise. It probably should be higher as they are more important. @benthorner do you have thoughts on that? It doesn't really have an impact as we don't expect that queue to grow big.

Even with them being the same priority it can't be blocked by send_email_immediate_high. The priority indicates how frequently the queue is checked so each time Sidekiq needs to find a new job it'll check one of it's queues, and the priority indicates that it'll check send_email_transactional 8 times more frequently than the default queue.  In practice this means that means that even with a huge send_email_immediate_high queue to work through the first items in every other queue will still get worked on rapidly.

The situation that could block this is if all the Sidekiq workers were busy with a slow running task. For example if we had 300 content changes to process, which take 30 mins to process, and they exhausted all workers then we could find all email is blocked. If we were worried about that scenario we'd probably want to run a separate Sidekiq process for email sending.

Reviewer 2: Ah, thank you for the detailed explanation and for clearing up my concern! 

I agree it feels like transactional emails should be deemed as a higher priority even if the priority isn't going to make much of a difference in practice, due to as you say, there is only going to be a handful of them in the queue at any one time.

Reviewer: Thanks for spotting @1pretz1.

I don't mind much either way. An argument for keeping it as-is would be that it emphasises how the intuitive interpretation of "priority" does not match the way Sidekiq works. As long as they are processed more quickly than most other email, I think it's fine to leave it as-is, but I also don't mind if we increase it symbolically.

### Monitoring Email Alert Service

Email Alert Service listens for [Publishing API][] events and communicates
these to Email Alert API. We currently [only have alerts][rabbitmq-alert] that
occur in-hours if this system fails, so there is no visibility for any
out-of-hours problems.

We intend to improve and consolidate the existing alerts, and contact an
engineer out-of-hours if there is a pattern of failure in processing these
events.

### Monitoring Email Alert Frontend

Email Alert Frontend acts as the public interface for subscriber interaction
with the email alert system. This web application will be monitored in a
[similar way to the Email Alert API web application](#web-application) by
considering whether the load balancer has any healthy instances.

### Removal of Email Alert Monitoring

[Email Alert Monitoring][email-alert-monitoring] is the aforementioned
end-to-end check that monitors whether travel advice and medical safety content
changes result in an email received by a Gmail inbox. We intend to retire this
check as we believe the new approach to monitoring the email alert system will
cover the aspects of this check that are within scope of the email alert system.

> Broadly the Email Alert Monitoring verifies that:

Reviewer 2: ```suggestion
Broadly Email Alert Monitoring verifies that:
```
